src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -1808,6 +1956,105 @@ open_common(int (*real_open)(const char *pathname, int oflags, ...), const char
 	if (!is_target_path)
 		goto org_func;
 
+	/* Always rely on fuse for open() to avoid a fake fd */
+	if (compatible_mode) {
+		dfs_obj_t    *dfs_obj_local;
+		struct ht_fd *fd_ht_obj = NULL;
+		int           fd_fake;
+
+		if (two_args)
+			fd_kernel = real_open(pathname, oflags);
+		else
+			fd_kernel = real_open(pathname, oflags, mode);
+		/* dfuse fails for some reason */
+		if (fd_kernel < 0)
+			goto out_compatible;
+		/* query file object through ioctl() */
+		rc = fetch_dfs_cont_file_obj_with_fd(fd_kernel, dfs_mt, &dfs_obj_local);
+		if (rc)
+			goto out_compatible;
+		/* Need to create a fake fd and associate with fd_kernel */
+		atomic_fetch_add_relaxed(&num_open, 1);
+
+		rc = dfs_get_mode(dfs_obj_local, &mode_query);
+		if (rc)
+			goto out_compatible;
+
+		if (S_ISREG(mode_query)) {
+			/* regular file */
+			rc = find_next_available_fd(NULL, &idx_fd);
+			if (rc)
+				goto out_compatible;
+			file_list[idx_fd]->dfs_mt    = dfs_mt;
+			file_list[idx_fd]->file      = dfs_obj_local;
+			file_list[idx_fd]->parent    = parent;
+			file_list[idx_fd]->st_ino    = FAKE_ST_INO(full_path);
+			file_list[idx_fd]->idx_mmap  = -1;
+			file_list[idx_fd]->open_flag = oflags;
+			file_list[idx_fd]->offset    = 0;
+			strncpy(file_list[idx_fd]->item_name, item_name, DFS_MAX_NAME);
+			fd_fake = idx_fd + FD_FILE_BASE;
+		} else if (S_ISDIR(mode_query)) {
+			/* directory */
+			rc = find_next_available_dirfd(NULL, &idx_dirfd);
+			if (rc)
+				goto out_compatible;
+			dir_list[idx_dirfd]->dfs_mt   = dfs_mt;
+			dir_list[idx_dirfd]->fd       = idx_dirfd + FD_DIR_BASE;
+			dir_list[idx_dirfd]->offset   = 0;
+			dir_list[idx_dirfd]->dir      = dfs_obj_local;
+			dir_list[idx_dirfd]->num_ents = 0;
+			dir_list[idx_dirfd]->st_ino   = FAKE_ST_INO(full_path);
+			memset(&dir_list[idx_dirfd]->anchor, 0, sizeof(daos_anchor_t));
+			dir_list[idx_dirfd]->path = NULL;
+			dir_list[idx_dirfd]->ents = NULL;
+			if (strncmp(full_path, "/", 2) == 0)
+				full_path[0] = 0;
+
+			/* allocate memory for ents[]. */
+			D_ALLOC_ARRAY(dir_list[idx_dirfd]->ents, READ_DIR_BATCH_SIZE);
+			if (dir_list[idx_dirfd]->ents == NULL) {
+				free_dirfd(idx_dirfd);
+				goto out_compatible;
+			}
+			D_ASPRINTF(dir_list[idx_dirfd]->path, "%s%s", dfs_mt->fs_root, full_path);
+			if (dir_list[idx_dirfd]->path == NULL) {
+				free_dirfd(idx_dirfd);
+				goto out_compatible;
```

Comment:
free dir_list[idx_dirfd]->ents ?

Reply:
"dir_list[idx_dirfd]->ents" and "dir_list[idx_dirfd]->path" were freed inside free_dirfd(idx_dirfd).